### PR TITLE
Change penalty rules. Small optimization. Add optional argument mask_pattern.

### LINKED ISSUE #125: Penalty rule 1 seems incorrect.
_lost_point_level1() calculates the total number of a pixel's adjacent pixels whose color is the same as that pixel, by iterate through row_range (-1, 0, 1) and col_range (-1, 0, 1).
e.g. in a 3x3 sub matrix and center pixel color is black, count the number of black pixels in 8
border pixels. And then use later for the penalty point count.
But I think That's incorrect. Rule 1 should only rule rows, or columns.
i.e. 5 consecutive color in a row: drop 3 points. 6 consecutive color in a column: drop 4 points.
Here's the rule description in *ISOIEC 18004_2006Cor 1_2009.pdf* [download link](https://qrcodenet.codeplex.com/):

|Feature|Evaluation condition|Points|
|---------|---------------|-------------------|
|Adjacent modules in row/column in same color|No. of modules = (5 + i)|N 1 + i|
|Block of modules in same color|Block size = m × n|N 2 × (m - 1) × (n - 1)|

Rule 2 is for block. So rule 1 won't rule block. Rule 1 should just rule a row, or a column.
Below all agree with the second interpretation and implementation:
1. `IDFS_2012_QR code_suplement.pdf` download link same as above.
2. [Thonky's tutorial](http://www.thonky.com/qr-code-tutorial/data-masking) site contains the same picture as 1's pdf, maybe from the same author?
3. [Zebra crossing](https://github.com/zxing/zxing/blob/master/core/src/main/java/com/google/zxing/qrcode/encoder/MaskUtil.java#L37-L42): Popular Java qrcode generator that has 10,000+ stars.

Maybe I'll file a pull request for this if y'all don't mind and if my implementation is correct and pass the correctness test.
By the way, is the algorithm lower bound O(n\*n) for iterating through a 2d matrix?
Is there any algorithm better than O(n\*n) for this penalty rule 1?
Best regards.

### LINKED ISSUE #126: Allow user to skip the best_mask_pattern procedure
Since the low performance described in #124 , it's better to allow user to skip the best_mask_pattern procedure if he/she knows the consequence and is willing to take that risk. My argument is as follows:
1. Faster. Skip finding best mask pattern can reduce 2/3 of total time generating a qrcode. Test:
```
import qrcode
import time

def make():
    qr = qrcode.QRCode(
        version=6,
        error_correction=qrcode.constants.ERROR_CORRECT_H,
        box_size=10,
        border=4,
    )
    qr.add_data('HELLO WORLDHELLO WORLDHELLO WORLDHELLO WORLDHELLO WORLDHELLO WORLDHELLO WORLDHELLO WORLDHELLO WORLDHELLO WORLDHELLO WORLDHELLO')
    qr.make(fit=True)

    img = qr.make_image()

start = time.time()
for i in range(100):
    make()
end = time.time()
print(end-start)
```
output: 14.734620332717896
And if I comment out [these lines](https://github.com/lincolnloop/python-qrcode/blob/master/qrcode/main.py#L162-L169) to skip the finding of best mask pattern,
output: 5.343664646148682
PS: Using different [penalty rule 1](#125) implementation I wrote.

2. Many other (informal) qrcode generators don't implement any penalty rules at all. And some are super fast.
Like this: [goqr](https://github.com/RaymondChou/goqr), [English fork](https://github.com/loggi/goqr). Written in Go language. The author said that it can *generates 1000 qrcodes in 8s, 8ms/image on everage. Tested under MacBook Pro / i5 2.5G / 4G / Go 1.0.1.* Pretty fast.

3. And if using in Django for generating 50+ qrcodes in one page for a user, CPU will reach 100% during that period(30+ seconds). This is good for DDoS attack.
Best wishes.

### PR DESCRIPTION
**Pull request by cryptogun**
Fixes #125 . Fixes #126 .

### CODE REVIEW

File: qrcode/main.py [hunk 1]
```diff
@@ -23,11 +23,22 @@ def _check_box_size(size):
...
+    if mask_pattern not in range(8):
+        raise ValueError(
+            "Mask pattern should be in range(8) (got %s)" % mask_pattern)
+
 class QRCode:
 
     def __init__(self, version=None,
                  error_correction=constants.ERROR_CORRECT_M,
                  box_size=10, border=4,
+                 mask_pattern=None,
```
Comment (SmileyChris) [inline qrcode/main.py]: Best to add new arguments to the end so it won't break if anyone is only using positional arguments.

File: qrcode/main.py [hunk 2]
```diff
@@ -76,7 +89,10 @@ def make(self, fit=True):
         """
         if fit or (self.version is None):
             self.best_fit(start=self.version)
-        self.makeImpl(False, self.best_mask_pattern())
+        if self.mask_pattern == None:
```
Comment (SmileyChris) [inline qrcode/main.py]: `is None` (or you could just do a single `self.mask_pattern if self.mask_pattern else self.best_mask_pattern()`)

**Comment by SmileyChris** [comment]
Nice work! I'll take a closer review of this early next year and we can roll out a release then :)

File: qrcode/main.py (hunk 2 above)
Comment (cryptogun) [inline qrcode/main.py]: Right. Now I learn that it's better to use `is` [when you want to check against an object's identity](https://stackoverflow.com/questions/14247373/python-none-comparison-should-i-use-is-or).
self.mask_pattern can be 0 so the latter should changed to `self.mask_pattern if self.mask_pattern is None else self.best_mask_pattern()`. And that may exceed 80 characters in one line.

File: qrcode/util.py [hunk 3]
```diff
@@ -189,51 +189,40 @@ def _lost_point_level1(modules, modules_count):
...
-                        sameCount += 1
+                if length >= 5:
+                    container[length] += 1
+                length = 1
+                previous_color = this_row[col]
+        if length >= 5:
+            container[length] += 1
+
+    for col in modules_range:
+        #this_col = [this_row[col] for this_row in modules] Slow, commented out.
```
Comment (cryptogun) [inline qrcode/util.py]: If so, I'll drop it out.
Moving `this_row` out of the inner for loop can increase the speed a bit. But `this_col` will be slower. I did that to leave a message so others won't move `this_col` out again for optimization.

**Comment by SmileyChris** [comment]
Thanks @cryptogun for your work here. Apologies for how long it took to get merged :sleeping:

**Comment by cryptogun** [comment]
@SmileyChris you're welcome~ I'm busy doing other things too.
P.S: Maybe you should remove the last commit (i.e. pip install ... from my own GitHub url). (Seems that you've done so.)

**Comment by SmileyChris** [comment]
Yeah, I did it as part of the merge commit. Thanks again!

**Pull request merged**